src/dune_engine/process.ml

```diff
@@ -74,20 +74,23 @@ module Io = struct
     ; mutable status : status
     }
 
-  let terminal ch =
+  let terminal ch ~swallow_on_success =
     let fd = descr_of_channel ch in
-    { kind = Terminal
+    { kind = Terminal { swallow_on_success }
     ; mode = mode_of_channel ch
     ; fd = lazy fd
     ; channel = lazy ch
     ; status = Keep_open
     }
 
-  let stdout = terminal (Out_chan stdout)
+  let stdout_swallow_on_success =
+    terminal (Out_chan stdout) ~swallow_on_success:true
 
-  let stderr = terminal (Out_chan stderr)
+  let stdout = terminal (Out_chan stdout) ~swallow_on_success:false
 
-  let stdin = terminal (In_chan stdin)
+  let stderr = terminal (Out_chan stderr) ~swallow_on_success:false
+
+  let stdin = terminal (In_chan stdin) ~swallow_on_success:false
```

Comment:
It seems weird to pass a meaningless value of `swallow_on_success` here. Perhaps, it should actually be a field of `Out_chan` constructor?

Reply:
(Actually, adding it to `Out_chan` doesn't seem to make sense. And using GADT feels too heavy here.)

Reply:
Agreed. I'll add a comment about this.

Reply:
I added a comment